File: sdk/runtime-client.py
"""Thin standard-library client for the DungeonQ reference HTTP API.

No automatic retries. Keep the same operation requestId after an uncertain
response, and inspect server evidence before deciding how to continue.
"""
import json
import re
import urllib.error
import urllib.parse
import urllib.request


class RuntimeError(Exception):
    def __init__(self, code, status=0, request_id=None, uncertain=False):
        super().__init__(code)
        self.code = code
        self.status = status
        self.request_id = request_id
        self.uncertain = uncertain


class _NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


class RuntimeClient:
    def __init__(self, origin, token="", timeout=10):
        try:
            url = urllib.parse.urlsplit(origin)
            url.port  # Validate the port before a request can begin.
        except (ValueError, TypeError, AttributeError):
            raise RuntimeError("INVALID_RUNTIME_ORIGIN") from None
        if (url.scheme not in ("http", "https") or not url.hostname
                or (url.scheme == "http" and url.hostname not in ("127.0.0.1", "localhost", "::1"))
                or url.username or url.password or url.query or url.fragment or url.path not in ("", "/")):
            raise RuntimeError("INVALID_RUNTIME_ORIGIN")
        if not isinstance(timeout, (int, float)) or isinstance(timeout, bool) or not 0 < timeout <= 60:
            raise RuntimeError("INVALID_CLIENT_OPTIONS")
        self._origin = urllib.parse.urlunsplit((url.scheme, url.netloc, "", "", ""))
        self._timeout = timeout
        self._opener = urllib.request.build_opener(_NoRedirect(), urllib.request.ProxyHandler({}))
        self.set_token(token)

    def set_token(self, token):
        if not isinstance(token, str) or len(token) > 4096 or re.search(r"[\s\x00-\x1f\x7f]", token):
            raise RuntimeError("INVALID_TOKEN")
        self._token = token

    def disconnect(self):
        self._token = ""

    def _request(self, path, body=None, authenticated=True):
        if authenticated and not self._token:
            raise RuntimeError("AUTH_REQUIRED")
        headers = {"Accept": "application/json", "Cache-Control": "no-store"}
        if authenticated:
            headers["Authorization"] = "Bearer " + self._token
        data = None
        if body is not None:
            headers["Content-Type"] = "application/json"
            data = json.dumps(body, separators=(",", ":")).encode("utf-8")
        request = urllib.request.Request(self._origin + path, data=data, headers=headers,
                                         method="POST" if data is not None else "GET")
        try:
            try:
                response = self._opener.open(request, timeout=self._timeout)
            except urllib.error.HTTPError as error:
                response = error
            with response:
                status = response.code
                content = response.read(2_097_153)
            if len(content) > 2_097_152:
                raise RuntimeError("RESPONSE_TOO_LARGE", status, uncertain=body is not None)
            try:
                value = json.loads(content)
            except (ValueError, UnicodeError):
                raise RuntimeError("INVALID_RESPONSE", status, uncertain=body is not None) from None
            if not isinstance(value, dict):
                raise RuntimeError("INVALID_RESPONSE", status, uncertain=body is not None)
            if not 200 <= status < 300:
                code = value.get("error", {}).get("code") if isinstance(value.get("error"), dict) else None
                if not isinstance(code, str) or not re.fullmatch(r"[A-Z][A-Z0-9_]{0,95}", code):
                    code = "AUTH_REQUIRED" if status == 401 else "PERMISSION_DENIED" if status == 403 else "REQUEST_REJECTED"
                request_id = value.get("requestId")
                if not isinstance(request_id, str) or not re.fullmatch(r"[A-Za-z0-9_-]{1,128}", request_id):
                    request_id = None
                raise RuntimeError(code, status, request_id, uncertain=body is not None and status >= 500)
            return value
        except RuntimeError:
            raise
        except (urllib.error.URLError, TimeoutError, OSError):
            raise RuntimeError("CONNECTION_UNAVAILABLE", uncertain=body is not None) from None

    def capabilities(self):
        return self._request("/api/capabilities", authenticated=False)

    def status(self):
        return self._request("/api/status")

    def evidence(self):
        return self._request("/api/evidence")

    def preview(self, context_id, action, reason=None):
        body = {"contextId": context_id, "action": action}
        if reason is not None:
            body["reason"] = reason
        return self._request("/api/policy/preview", body)

    def apply(self, proposal_id, digest, confirmation):
        if confirmation != "APPLY":
            raise RuntimeError("EXPLICIT_CONFIRMATION_REQUIRED")
        return self._request("/api/policy/apply", {"proposalId": proposal_id, "digest": digest, "confirmation": confirmation})

    def operate(self, request_id, operation, args=None):
        if operation not in ("snapshot", "read", "write", "issue-ticket", "use-ticket"):
            raise RuntimeError("UNSUPPORTED_OPERATION")
        return self._request("/api/operate", {"requestId": request_id, "operation": operation, "args": {} if args is None else args})


def evidence_state(value):
    state = value.get("status") if isinstance(value, dict) else None
    checks = value.get("checks") if isinstance(value, dict) else None
    if state == "FAIL" or (isinstance(checks, list) and any(isinstance(item, dict) and item.get("status") == "FAIL" for item in checks)):
        return "FAIL"
    if state == "PASS" and "checks" in value and (not isinstance(checks, list) or not checks or any(not isinstance(item, dict) or item.get("status") != "PASS" for item in checks)):
        return "INCONCLUSIVE"
    return state if state in ("PASS", "FAIL", "INCONCLUSIVE") else "INCONCLUSIVE"

File: sdk/runtime-client.mjs
// Shared browser / Node transport. Authority and validation belong to the server.
export class RuntimeError extends Error {
  constructor(code, { status = 0, requestId, uncertain = false } = {}) {
    super(code);
    this.name = 'RuntimeError';
    this.code = code;
    this.status = status;
    this.requestId = requestId;
    this.uncertain = uncertain;
  }
}

function originOf(value) {
  let url;
  try { url = new URL(value); } catch { throw new RuntimeError('INVALID_RUNTIME_ORIGIN'); }
  const loopback = ['127.0.0.1', 'localhost', '[::1]'].includes(url.hostname);
  if (!['https:', 'http:'].includes(url.protocol) || (url.protocol === 'http:' && !loopback)
    || url.username || url.password || url.search || url.hash || !['', '/'].includes(url.pathname)) {
    throw new RuntimeError('INVALID_RUNTIME_ORIGIN');
  }
  return url.origin;
}

const operations = new Set(['snapshot', 'read', 'write', 'issue-ticket', 'use-ticket']);
const safeCode = value => typeof value === 'string' && /^[A-Z][A-Z0-9_]{0,95}$/.test(value) ? value : null;
async function boundedText(response, uncertain) {
  if (!response.body) return '';
  const reader = response.body.getReader(); const decoder = new TextDecoder();
  let size = 0; let content = '';
  try {
    for (;;) {
      const next = await reader.read();
      if (next.done) break;
      size += next.value.byteLength;
      if (size > 2_097_152) throw new RuntimeError('RESPONSE_TOO_LARGE', { status: response.status, uncertain });
      content += decoder.decode(next.value, { stream: true });
    }
    return content + decoder.decode();
  } finally { await reader.cancel().catch(() => {}); }
}

export function createRuntimeClient({ origin, token = '', timeoutMs = 10000, fetch: fetcher = globalThis.fetch } = {}) {
  const base = originOf(origin);
  if (typeof fetcher !== 'function' || !Number.isInteger(timeoutMs) || timeoutMs < 1 || timeoutMs > 60000) {
    throw new RuntimeError('INVALID_CLIENT_OPTIONS');
  }
  let credential = '';
  let generation = 0;
  const pending = new Set();
  function setToken(value) {
    if (typeof value !== 'string' || value.length > 4096 || /[\s\x00-\x1f\x7f]/.test(value)) throw new RuntimeError('INVALID_TOKEN');
    generation++;
    for (const controller of pending) controller.abort();
    credential = value;
  }
  setToken(token);
  async function request(path, body, authenticated = true) {
    if (authenticated && !credential) throw new RuntimeError('AUTH_REQUIRED');
    const atGeneration = generation;
    const controller = new AbortController();
    pending.add(controller);
    const timer = setTimeout(() => controller.abort(), timeoutMs);
    const mutating = body !== undefined;
    try {
      const response = await fetcher(base + path, {
        method: mutating ? 'POST' : 'GET',
        headers: { Accept: 'application/json', ...(authenticated ? { Authorization: `Bearer ${credential}` } : {}),
          ...(mutating ? { 'Content-Type': 'application/json' } : {}) },
        ...(mutating ? { body: JSON.stringify(body) } : {}),
        credentials: 'omit', cache: 'no-store', redirect: 'error', signal: controller.signal,
      });
      if (generation !== atGeneration) throw new RuntimeError('SESSION_CHANGED', { uncertain: mutating });
      const content = await boundedText(response, mutating);
      if (generation !== atGeneration) throw new RuntimeError('SESSION_CHANGED', { uncertain: mutating });
      let value;
      try { value = JSON.parse(content); } catch {
        throw new RuntimeError('INVALID_RESPONSE', { status: response.status, uncertain: mutating });
      }
      if (!value || typeof value !== 'object' || Array.isArray(value)) throw new RuntimeError('INVALID_RESPONSE', { uncertain: mutating });
      if (!response.ok) {
        const code = safeCode(value.error?.code) ?? (response.status === 401 ? 'AUTH_REQUIRED'
          : response.status === 403 ? 'PERMISSION_DENIED' : 'REQUEST_REJECTED');
        throw new RuntimeError(code, { status: response.status, uncertain: mutating && response.status >= 500,
          requestId: typeof value.requestId === 'string' && /^[A-Za-z0-9_-]{1,128}$/.test(value.requestId) ? value.requestId : undefined });
      }
      return value;
    } catch (error) {
      if (error instanceof RuntimeError) throw error;
      throw new RuntimeError(generation !== atGeneration ? 'SESSION_CHANGED'
        : controller.signal.aborted ? 'REQUEST_TIMEOUT' : 'CONNECTION_UNAVAILABLE', { uncertain: mutating });
    } finally { clearTimeout(timer); pending.delete(controller); }
  }
  return Object.freeze({
    setToken,
    disconnect() { setToken(''); },
    capabilities: () => request('/api/capabilities', undefined, false),
    status: () => request('/api/status'),
    evidence: () => request('/api/evidence'),
    preview(input) { return request('/api/policy/preview', input); },
    apply({ proposalId, digest, confirmation } = {}) {
      if (confirmation !== 'APPLY') throw new RuntimeError('EXPLICIT_CONFIRMATION_REQUIRED');
      return request('/api/policy/apply', { proposalId, digest, confirmation });
    },
    operate({ requestId, operation, args = {} } = {}) {
      if (!operations.has(operation)) throw new RuntimeError('UNSUPPORTED_OPERATION');
      return request('/api/operate', { requestId, operation, args });
    },
  });
}

// Missing or unfamiliar result states are never promoted to a successful check.
export function evidenceState(value) {
  const state = ['PASS', 'FAIL', 'INCONCLUSIVE'].includes(value?.status) ? value.status : 'INCONCLUSIVE';
  const checks = Array.isArray(value?.checks) ? value.checks : null;
  if (state === 'FAIL' || checks?.some(check => check?.status === 'FAIL')) return 'FAIL';
  if (state === 'PASS' && Object.hasOwn(value, 'checks')
    && (!checks || !checks.length || checks.some(check => check?.status !== 'PASS'))) return 'INCONCLUSIVE';
  return state;
}
